Src/Particle/AMReX_ParticleUtil.H

```diff
@@ -758,6 +758,126 @@ void clearEmptyEntries (C& c)
+
+    Gpu::DeviceVector<index_type> llist_start(nbins, llist_guard);
+    Gpu::DeviceVector<index_type> llist_next(nitems);
+    perm.resize(nitems);
+    Gpu::DeviceScalar<index_type> global_idx(0);
+
+    index_type* pllist_start = llist_start.dataPtr();
+    index_type* pllist_next = llist_next.dataPtr();
+    index_type* pperm = perm.dataPtr();
+    index_type* pglobal_idx = global_idx.dataPtr();
+
+    amrex::ParallelFor(nitems, [=] AMREX_GPU_DEVICE (index_type i) noexcept
+    {
+        i = nitems - i - 1;
+        pllist_next[i] = Gpu::Atomic::Exch(pllist_start + f(i), i);
+    });
+
+#if defined(AMREX_USE_CUDA) || defined(AMREX_USE_HIP)
+    amrex::launch<gpu_block_size>(nbins / gpu_block_size, Gpu::gpuStream(),
+        [=] AMREX_GPU_DEVICE () {
+            __shared__ index_type sdata[gpu_block_size];
+            index_type current_idx = pllist_start[threadIdx.x + gpu_block_size * blockIdx.x];
+
+            while (true) {
+                sdata[threadIdx.x] = index_type(current_idx != llist_guard);
+                index_type x = 0;
+
+                // simple block wide prefix sum
+                for (index_type i = 1; i<gpu_block_size; i*=2) {
+                    __syncthreads();
+                    if (threadIdx.x >= i) {
+                        x = sdata[threadIdx.x - i];
+                    }
+                    __syncthreads();
+                    if (threadIdx.x >= i) {
+                        sdata[threadIdx.x] += x;
+                    }
+                }
+                __syncthreads();
+                if (sdata[gpu_block_size_m1] == 0) {
+                    break;
+                }
+                __syncthreads();
+                if (threadIdx.x == gpu_block_size_m1) {
+                    x = sdata[gpu_block_size_m1];
+                    sdata[gpu_block_size_m1] = Gpu::Atomic::Add(pglobal_idx, x);
+                }
+                __syncthreads();
+                if (threadIdx.x < gpu_block_size_m1) {
+                    sdata[threadIdx.x] += sdata[gpu_block_size_m1];
+                }
+                __syncthreads();
+                if (threadIdx.x == gpu_block_size_m1) {
+                    sdata[gpu_block_size_m1] += x;
+                }
+                __syncthreads();
+
+                if (current_idx != llist_guard) {
+                    pperm[sdata[threadIdx.x] - 1] = current_idx;
+                    current_idx = pllist_next[current_idx];
+                }
+            }
+        });
+#else
+    Abort("Not implemented");
+#endif
+
+    Gpu::Device::streamSynchronize();
+}
+
+template <class index_type, class pos_struct>
+void PermutationForDeposition (Gpu::DeviceVector<index_type>& perm, index_type nitems,
+                               pos_struct const* v, Box bx, Geometry geom, const IntVect idx_type)
+{
+    AMREX_ALWAYS_ASSERT(idx_type.allGE(IntVect(0)) && idx_type.allLE(IntVect(2)));
+
+    const IntVect refine_vect = max(idx_type, IntVect(1)).min(IntVect(2));
```

Comment:
This may not be correct. It's doing something like this
```
IntVect tmp = amrex:max(idx_type, IntVect(1)); // This is a lexicographical comparison
const IntVect refine_vect = tmp.min(IntVect(2)); // This is an element wise operation.
```

Reply:
I want to use this function so idx_type does not change.
```
/**
 * \brief Returns the IntVect that is the component-wise maximum of two
 * argument IntVects.
 */
AMREX_GPU_HOST_DEVICE
AMREX_FORCE_INLINE
IntVect
max (const IntVect& p1, const IntVect& p2) noexcept
{
    IntVect p(p1);
    p.max(p2);
    return p;
}
```
That should be element wise

Reply:
Okay.  Forgot we have a specialization for IntVect.  I thought it will pick up the function in AMReX_Algorithm.H.